ARCHITECTURE.md

```diff
@@ -0,0 +1,131 @@
+# OpenSearch Security Plugin Architecture
+
+OpenSearch’s core systems do not include security features, these features are added by installing the Security Plugin.  The Security Plugin accesses OpenSearch’s internals to build up Authentication, Authorization, End to End Encryption, Audit Logging, and management interfaces.
+
+## Components
+
+The Security Plugin is packaged into a standard plugin zip file used by OpenSearch which can be installed by using the plugin tool.  The security configuration is accessible on disk for modification before the node has been turned on, afterward the admin tools or api endpoints need to be used.
+
+```mermaid
+graph TD
+    subgraph OpenSearch Node
+        subgraph File System
+            cfg[Security Configuration files]
+            adm[Admin Tools]
+        end
+        subgraph Indices
+            idx(Index 1..n)
+            secIdx[Security Index]
+        end
+        subgraph Plugins
+           pgns(Plugins 1..n)
+           sec[Security Plugin]
+        end
+
+        sec -- bootstrap security config --> cfg
+        sec -- refresh security config from cluster --> secIdx
+        adm -- backup/restore security config --> sec
+    end
+```
+
+### Security Plugin
+
+The runtime of the security plugin uses extension points to insert itself into the path actions.  Several security managemenet actions are registered in OpenSearch so they can be changed through REST API actions.
+
+### Security Configuration
+
+The security configuration is stored in an system index that is replicated to all nodes.  When a change has been made the Security plugin is reloaded to cleanly initialize its components with the new configuration.
+
+#### Configuration Files
+
+When starting up with no security index detected in the cluster the security plugin will attempt to load configuration files from disk into a new security index, these can be manually modifed or sourced from a backup of a security index from the admin tools.
+
+### Admin Tools
+
+For OpenSearch nodes to join a cluster they need to have the same security configuration.  During this joining period configuration values need to be consistant for SSL settings as well as certificate files need to be available on the node before it joins the cluster.  The admin tools ecapsulate these and other features.
+
+## Flows
+
+### Authentication / Authorization
+
+The Security plugin supports multiple authentication backends, there is an internal identity provider that works with http basic auth as well as support external providers including OpenId Connect (OIDC) and SAML.
+
+Authorization is governed by roles declared in the security configuration, roles control access to resource by the action name and/or index names in combination with OpenSearch action names.  These actions names are from the transport action name in OpenSearch.

```

Comment:
This is typically the case that transport action names are what is put in a role definition, but recently with the restapi: addition to the security plugin and rest layer authz it would also include named routes.

Reply:
A separate section around special permissions could be useful to capture details like this; I plan to keep the high level architecture simple.